File: 0082.remove_duplicates_from_sorted_list_II/solution.go
package leetcode0082

import (
	"github.com/Ackerr/Algorithms/utils"
)

func deleteDuplicates(head *utils.ListNode) *utils.ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	if head.Val == head.Next.Val {
		for head.Next != nil && head.Val == head.Next.Val {
			head = head.Next
		}
		return deleteDuplicates(head.Next)
	}
	head.Next = deleteDuplicates(head.Next)
	return head
}

File: 0082.remove_duplicates_from_sorted_list_II/solution.py
from utils import ListNode


class Solution:
    def delete_duplicates(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        node = ListNode(None)
        node.next = head
        target = node
        while target.next:
            left = right = target.next
            while right.next and right.next.val == left.val:  # duplicate
                right = right.next
            if left == right:  # no move
                target = target.next
            else:
                target.next = right.next  # remove duplicate node
        return node.next
